### PR TITLE
Fix possibility of deadlock in the SIGSEGV handler

### PR DESCRIPTION
**Pull request by janvorli**
The SIGSEGV handler was transitively calling pthread_attr_get_np to
get current stack limit no matter where the exception happened. The
problem was that if the segmentation violation happened in a C
runtime function, this could lead to deadlock since pthread_attr_get_np
is not signal safe.

The fix is to add callback that will figure out if the failing instruction
was in managed code or in one of the jit helpers and if not, it would
not attempt to handle the signal.

I have also removed non signal safe calls from the signal handlers code
path where we don't know yet whether the interrupted code was in a function
that could collide with a function that we call from the signal handler
or not. This includes TRACE and ASSERT calls that were calling string
formatting functions and modification of the TwoWayPipe to store the 
pipe name so that it doesn't have to regenerate it in TwoWayPipe::Disconnect
using signal unsafe functions.

In addition, I have taken the opportunity and unified the code of all the hardware
signal handlers, since each of them contained a copy of the same code. I have
moved the common code into the common_signal_handler.

### CODE REVIEW

**Comment by janvorli** [comment]
@jkotas can you take a look please?

**Comment by janvorli** [comment]
Fixes #3912

File: src/vm/exceptionhandling.cpp [hunk 1]
```diff
@@ -5031,65 +5032,67 @@ bool IsDivByZeroAnIntegerOverflow(PCONTEXT pContext)
 }
 #endif //_AMD64_
 
+BOOL PALAPI IsSafeToHandleHardwareException(PCONTEXT contextRecord, PEXCEPTION_RECORD exceptionRecord)
+{
+    PCODE controlPc = GetIP(contextRecord);
+    return g_fEEStarted && (
+        exceptionRecord->ExceptionCode == STATUS_BREAKPOINT || 
+        exceptionRecord->ExceptionCode == STATUS_SINGLE_STEP ||
+        ExecutionManager::IsManagedCode(controlPc) ||
```
Comment (jkotas) [inline src/vm/exceptionhandling.cpp]: `ExecutionManager::IsManagedCode` will take a lock unless we are in cooperative mode already. Do we need to check whether we are in cooperative mode before calling it?

**Comment by janvorli** [comment]
@jkotas, I've fixed the problem with potential deadlock and also made the same fix in the CheckActivationSafePoint.

**Comment by jkotas** [comment]
LGTM

**Pull request merged**